sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageTrigger.scala

```diff
@@ -0,0 +1,141 @@
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.util.{EventLoop, ThreadUtils}
+
+/**
+ * This class triggers [[QueryStage]] bottom-up, apply planner rules for query stages and
+ * materialize them. It triggers as many query stages as possible at the same time, and triggers
+ * the parent query stage when all its child stages are materialized.
+ */
+class QueryStageTrigger(session: SparkSession, callback: QueryStageTriggerCallback)
+  extends EventLoop[QueryStageTriggerEvent]("QueryStageTrigger") {
+
+  private val stageToParentStage = HashMap.empty[Int, QueryStage]
+
+  private val idToUpdatedStage = HashMap.empty[Int, QueryStage]
+
+  private val stageToNumPendingChildStages = HashMap.empty[Int, Int]
+
+  private val submittedStages = HashSet.empty[Int]
+
+  private val planner = new QueryStagePlanner(session.sessionState.conf)
+
+  def trigger(stage: QueryStage): Unit = {
+    // Avoid submitting reused QueryStage.
+    if (!submittedStages.contains(stage.id)) {
+      submittedStages += stage.id
+
+      val childStages = stage.plan.collect {
+        case stage: QueryStage => stage
+      }
+
+      if (childStages.isEmpty) {
+        // this is the leaf stage, submit it
+        post(SubmitStage(stage))
+      } else {
+        // We don't submit duplicated stages, here we deduplicate the childStages first.
+        val deduped = childStages.groupBy(_.id).values.map(_.head).toSeq
+        stageToNumPendingChildStages(stage.id) = deduped.length
+        // trigger child stages first
+        deduped.foreach { child =>
+          stageToParentStage(child.id) = stage
+          trigger(child)
+        }
+      }
+    }
+  }
+
+  private implicit def executionContext: ExecutionContextExecutorService = {
+    QueryStageTrigger.executionContext
+  }
+
+  override protected def onReceive(event: QueryStageTriggerEvent): Unit = event match {
+    case SubmitStage(stage) =>
+      Future {
+        // planning needs active SparkSession in current thread.
+        SparkSession.setActiveSession(session)
+        planner.execute(stage.plan)
+      }.onComplete { res =>
+        if (res.isSuccess) {
+          post(StagePlanned(stage, res.get))
+        } else {
+          callback.onStagePlanningFailed(stage, res.failed.get)
+          stop()
+        }
+      }
+
+    case StagePlanned(stage, optimizedPlan) =>
+      val newStage = stage.withNewPlan(optimizedPlan)
+      idToUpdatedStage(newStage.id) = newStage
+      // This stage has optimized its plan, notify the callback about this change.
+      callback.onStageUpdated(newStage)
+
+      newStage.materialize().onComplete { res =>
+        if (res.isSuccess) {
+          post(StageReady(stage))
+        } else {
+          callback.onStageMaterializingFailed(newStage, res.failed.get)
+          stop()
+        }
+      }
+
+    case StageReady(stage) =>
+      stageToParentStage.remove(stage.id).foreach { parentStage =>
```

Comment:
A stage being reused can have multiple parent stages. Need decrease all parent stages' numPendingChildStages.

Reply:
It is also possible when the reused stage is ready, another parent have not yet been triggered, right? So before triggering the child stages, we may also check if any one has already been ready.